### PR TITLE
soc: riscv: riscv-privilege: telink_b91: dts: riscv: telink: telink_b…

### PR DESCRIPTION
**Pull request by andriy-bilynskyy**
…91: drivers: serial: uart_b91.c: Implement Telink B91 suspend mode

To reduce power consumption Telink B91 supports suspend mode.
This mode is involved when OS kernel enter idle process.

Co-authored-by: Andrii Bilynskyi <andrii.bilynskyi@telink-semi.com>

Signed-off-by: Misha Tkachenko <misha.tkachenko@telink-semi.com>

Tested manually using TLSR9518ADK80D by running sample projects with config PM.

### CODE REVIEW

**Review by s07641069 — DISMISSED** [review]

**Review by galak — DISMISSED** [review]
can you do the reformatting of the uart driver as its own commit.

**Comment by andriy-bilynskyy** [comment]
@galak formatting of uart driver is back only PM relative stuffs are added.

File: drivers/serial/uart_b91.c [hunk 1]
```diff
@@ -518,6 +522,36 @@ static void uart_b91_irq_callback_set(const struct device *dev,
...
+		/* reset TX/RX byte index */
+		data->tx_byte_index = 0;
+		data->rx_byte_index = 0;
+		uart->status |= UART_RX_RESET_BIT | UART_TX_RESET_BIT;
+		break;
+
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* wait for TX done before entering suspend mode */
+		while (!(uart->txrx_status & UART_TXRX_STATUS_TX_DONE)) {
+		};
```
Comment (gmarull) [inline drivers/serial/uart_b91.c:543]: please make use of `pm_policy_state_lock_get`/`pm_policy_state_lock_put` API, the PM callback is not a place where one should run busy loops. That means system is trying to enter a PM state that is not supported when UART is active.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_b91.c (hunk 1 above)
Comment (andriy-bilynskyy) [reply drivers/serial/uart_b91.c:543]: Entering suspend on Telink B91 affects all peripheral devices
It's not just core functionality as most other vendors.
It mean no way to separate CPU sleep and peripheral devices sleep. 
So if some device is busy with its operation it should prevent entering
suspend mode at all.
from zephyr/subsys/pm/pm.c
I see the only way to prevent enter suspend
is to use pm_suspend_devices();
in that case corresponding device should return error
or finish its operations (as it's implemented).
Returning an error produces kernel trace about no suspend entered.
So it's implemented by this way.
Suggested CONFIG_PM_DEVICE_RUNTIME disables preventing core suspend entering.
May be I missed something. If there is a better option please suggest it.

Comment (andriy-bilynskyy) [reply drivers/serial/uart_b91.c]: polling is removed from PM callback.

File: drivers/serial/uart_b91.c [hunk 2]
```diff
@@ -341,6 +345,9 @@ static void uart_b91_poll_out(const struct device *dev, uint8_t c)
 
 	uart->data_buf[data->tx_byte_index] = c;
 	data->tx_byte_index = (data->tx_byte_index + 1) % ARRAY_SIZE(uart->data_buf);
+
+	while (!(uart->txrx_status & UART_TXRX_STATUS_TX_DONE)) {
+	};
```
Comment (fkokosinski) [inline drivers/serial/uart_b91.c:350]: Unnecessary semicolon?

File: drivers/serial/uart_b91.c [hunk 3]
```diff
@@ -386,7 +394,11 @@ static int uart_b91_fifo_fill(const struct device *dev,
 			break;
 		}
 
-		uart_b91_poll_out(dev, tx_data[i]);
+		while (uart_b91_get_tx_bufcnt(uart) >= UART_TX_BUF_CNT) {
+		};
```
Comment (fkokosinski) [inline drivers/serial/uart_b91.c:398]: Ditto.

File: soc/riscv/riscv-privilege/telink_b91/power.c [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+	switch (state) {
+	case PM_STATE_SUSPEND_TO_IDLE:
+		/* Nothing to do. */
+		break;
+
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
```
Comment (fkokosinski) [inline soc/riscv/riscv-privilege/telink_b91/power.c:128]: So what do we are actually want to do here? Because right now this switch statement doesn't do anything other than logging when `state != PM_STATE_SUSPEND_TO_IDLE`.

Comment (andriy-bilynskyy) [reply soc/riscv/riscv-privilege/telink_b91/power.c:128]: yes sure just notification that power mode is not enterd.

File: drivers/serial/uart_b91.c (hunk 2 above)
Comment (andriy-bilynskyy) [reply drivers/serial/uart_b91.c:350]: Same file L343 (no changes applied)
```
	while (uart_b91_get_tx_bufcnt(uart) >= UART_TX_BUF_CNT) {
	};
```
file is already formatted by this way.

File: drivers/serial/uart_b91.c (hunk 3 above)
Comment (andriy-bilynskyy) [reply drivers/serial/uart_b91.c:398]: See previous

File: drivers/serial/uart_b91.c (hunk 2 above)
Comment (stephanosio) [reply drivers/serial/uart_b91.c:350]: both should be fixed then

Comment (andriy-bilynskyy) [reply drivers/serial/uart_b91.c]: fixed

**Review by andy-liu-telink — DISMISSED** [review]

**Comment by andriy-bilynskyy** [comment]
Could someone proceed with review this PR. Because dependent PR into hal telink is already merged and it block to introduce its functionality.

File: soc/riscv/riscv-privilege/telink_b91/Kconfig.defconfig.series [hunk 5]
```diff
@@ -53,4 +53,7 @@ config TEST_EXTRA_STACK_SIZE
 config HAS_FLASH_LOAD_OFFSET
 	default y if BOOTLOADER_MCUBOOT
 
+config LEGACY_INCLUDE_PATH
+	default y
```
Comment (gmarull) [inline soc/riscv/riscv-privilege/telink_b91/Kconfig.defconfig.series:57]: please remove that

**Comment by andy-liu-telink** [comment]
@gmarull  Are you okay with our latest code? If so, could you approve this pull request?

File: soc/riscv/riscv-privilege/telink_b91/power.c [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+#define DT_DRV_COMPAT telink_machine_timer
+
+#define MTIME_REG	DT_INST_REG_ADDR(0)
+#define MTIMECMP_REG	(DT_INST_REG_ADDR(0) + 8)
+
+#define mticks_to_systicks(mticks)                                                                 \
+	(((uint64_t)(mticks)*SYSTEM_TIMER_TICK_1S) / CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC)
+#define systicks_to_mticks(sticks)                                                                 \
+	(((uint64_t)(sticks)*CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC) / SYSTEM_TIMER_TICK_1S)
```
Comment (gmarull) [inline soc/riscv/riscv-privilege/telink_b91/power.c:24]: why isn't this using tick conversion utilities?

File: soc/riscv/riscv-privilege/telink_b91/power.c [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	ARG_UNUSED(substate_id);
+
+	uint32_t tl_sleep_tick = stimer_get_tick();
+	uint64_t current_time = get_mtime();
+	uint64_t wakeup_time = get_mtime_compare();
+
+	if (wakeup_time <= current_time) {
+		LOG_DBG("Sleep Time = 0 or less\n");
+		return;
+	}
```
Comment (gmarull) [inline soc/riscv/riscv-privilege/telink_b91/power.c:88]: can you please explain why this is dealing with machine timer registers here?

**Review by s07641069 — DISMISSED** [review]

File: soc/riscv/riscv-privilege/telink_b91/power.c [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+#ifdef CONFIG_PM_DEVICE
+	if (pm_device_is_any_busy()) {
+		return;
+	}
+#endif /* CONFIG_PM_DEVICE */
```
Comment (gmarull) [inline soc/riscv/riscv-privilege/telink_b91/power.c:93]: this is not where this check should be done. If your system can't go to sleep because a device activity doesn't allow for a particular power state you must use the policy state_lock APIs.

File: soc/riscv/riscv-privilege/telink_b91/power.c [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+	switch (state) {
+	case PM_STATE_SUSPEND_TO_IDLE:
+		/* Nothing to do. */
+		break;
+
+	default:
+		LOG_DBG("Unsupported power state %u", state);
+		break;
+	}
```
Comment (gmarull) [inline soc/riscv/riscv-privilege/telink_b91/power.c:134]: why is this needed if it's doing nothing?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: soc/riscv/riscv-privilege/telink_b91/Kconfig.defconfig.series (hunk 5 above)
Comment (andriy-bilynskyy) [reply soc/riscv/riscv-privilege/telink_b91/Kconfig.defconfig.series:57]: if do so for now our platform builds will fail. It's due to dependent PR in hal_telink where we have legacy paths.
dependent PR was created before removing this feature in main zephyr. so it's result of long PR process.
i'll remove it in next PR after fixing paths in hal_telink

Comment (gmarull) [reply soc/riscv/riscv-privilege/telink_b91/Kconfig.defconfig.series:57]: please open a PR fixing the issues instead, this can't go in.

Comment (gmarull) [reply soc/riscv/riscv-privilege/telink_b91/Kconfig.defconfig.series:57]: FYI, you may find this useful https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/utils/migrate_includes.py

File: soc/riscv/riscv-privilege/telink_b91/power.c (hunk 6 above)
Comment (andriy-bilynskyy) [reply soc/riscv/riscv-privilege/telink_b91/power.c:24]: systicks is not standard risc-v mtimer ticks. it's specific platform ticks which are used duting sleep mode.
so we are converting risc-v mtimer ticks (32K) to platform specific ticks (16M). it's not converting mtimer ticks to mS

File: soc/riscv/riscv-privilege/telink_b91/power.c (hunk 7 above)
Comment (andriy-bilynskyy) [reply soc/riscv/riscv-privilege/telink_b91/power.c:88]: our platfom stops MTIMER during sleep. so we should get mtimer value and restore after sleep

File: soc/riscv/riscv-privilege/telink_b91/power.c (hunk 8 above)
Comment (andriy-bilynskyy) [reply soc/riscv/riscv-privilege/telink_b91/power.c:93]: in our platform sleep stops all devices. so if any device is in process of its action we can't enter sllep. we cant separate core and devices because sleep mode affect everything.

File: soc/riscv/riscv-privilege/telink_b91/power.c (hunk 9 above)
Comment (andriy-bilynskyy) [reply soc/riscv/riscv-privilege/telink_b91/power.c:134]: it's providing traces. if its not acceptable i can remove it

File: soc/riscv/riscv-privilege/telink_b91/power.c (hunk 8 above)
Comment (gmarull) [reply soc/riscv/riscv-privilege/telink_b91/power.c:93]: so that means device drivers should lock sleep state/s while they are active.
